demos/common/cpp/models/src/image_model.cpp

```diff
@@ -15,24 +15,30 @@
 */
 
 #include "models/image_model.h"
-#include <utils/ocv_common.hpp>
 
-ImageModel::ImageModel(const std::string& modelFileName, bool useAutoResize) :
+ImageModel::ImageModel(const std::string& modelFileName, bool useAutoResize, const InputTransform& inputTransform) :
     ModelBase(modelFileName),
-    useAutoResize(useAutoResize) {
+    useAutoResize(useAutoResize),
+    inputTransform(inputTransform) {
 }
 
 std::shared_ptr<InternalModelData> ImageModel::preprocess(const InputData& inputData, InferenceEngine::InferRequest::Ptr& request) {
-    auto& img = inputData.asRef<ImageInputData>().inputImage;
+    const auto& origImg = inputData.asRef<ImageInputData>().inputImage;
+    const auto& img = inputTransform.call(origImg);
 
     if (useAutoResize) {
         /* Just set input blob containing read image. Resize and layout conversionx will be done automatically */
-        request->SetBlob(inputsNames[0], wrapMat2Blob(img));
+        request->SetBlob(inputsNames[0], wrapMat2Blob(img, !inputTransform.isTrivial));
         /* IE::Blob::Ptr from wrapMat2Blob() doesn't own data. Save the image to avoid deallocation before inference */
         return std::make_shared<InternalImageMatModelData>(img);
     }
     /* Resize and copy data from the image to the input blob */
     InferenceEngine::Blob::Ptr frameBlob = request->GetBlob(inputsNames[0]);
-    matU8ToBlob<uint8_t>(img, frameBlob);
+    if (inputTransform.isTrivial) {
+        matToBlob<uint8_t>(img, frameBlob);
+    }
+    else {
+        matToBlob<float_t>(img, frameBlob);
+    }
```

Comment:
Yes, if the model wrapper supports input normalization, the IE should set precision for input blob by condition, and it is done while calling `inputTransform.setPrecision()`.

Moving on to the next question, I am not sure we can hide this **if** because of 2 points:

1. We call `matToBlob()` with `unsigned char` type in mask_rcnn_demo, can we replace this type with `uint8_t`? If no, we should not convert this template function to ordinary one
2. In super_resolution wrapper we call `matToBlot<float_t>` function in preprocessing, but it seems the input has CV8U type, so in this case your assumption is not working

Reply:
My answers:
1. uint8_t is unsigned char (under the hood)
2. we should check super_resolution wrapper, possibly  there's an error cause if we provide data in CV8U, input precision should be uint8_t (unless there's error/functionality in IE that neutralizes this error)

Reply:
As we discussed in other thread, matToBlob function has to be refactored to avoid data loss in automatic typecasts

Reply:
The `MatToBlob()` is updated.